lumen/ai/coordinator.py

```diff
@@ -504,19 +525,38 @@ async def respond(self, messages: list[Message], **kwargs: dict[str, Any]) -> st
                 max_user_messages=self.history
             )
 
+            # Check if this is a follow-up question that can be answered with existing context
+            is_follow_up = await self._check_follow_up_question(messages)
+
             agents = {agent.name[:-5]: agent for agent in self.agents}
-            execution_graph = await self._compute_execution_graph(messages, agents)
+
+            # If it's a follow-up and we already have the data, we can skip some tools execution
+            if is_follow_up:
+                log_debug("\033[92mDetected follow-up question, using existing context\033[0m")
+                with self.interface.add_step(title="Using existing data context...", user="Assistant") as step:
+                    step.stream("Detected that this is a follow-up question related to the previous dataset.")
+                    step.stream("\n\nUsing the existing data in memory to answer without re-executing data retrieval.")
+                    step.success_title = "Using existing data for follow-up question"
+
+            execution_graph = await self._compute_execution_graph(messages, agents, is_follow_up=is_follow_up)
```

Comment:
We should have a stub on `Coordinator._compute_execution_graph` to define the API.